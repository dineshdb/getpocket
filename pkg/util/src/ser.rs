use serde::Serializer;

pub fn serialize_option_bool_as_int<S>(
    value: &Option<bool>,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    match value {
        Some(true) => serializer.serialize_i32(1),
        Some(false) | None => serializer.serialize_i32(0),
    }
}

pub fn serialize_vec_as_comma_separated<S>(vec: &[String], serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    let joined = vec.join(",");
    serializer.serialize_str(&joined)
}
