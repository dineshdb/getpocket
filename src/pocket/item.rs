use crate::der::bool_from_string;
use crate::der::i32_from_string;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct Item {
    pub item_id: String,
    pub resolved_id: String,
    pub given_url: String,
    pub given_title: String,
    #[serde(deserialize_with = "bool_from_string")]
    pub favorite: bool,
    pub status: ItemStatus,
    #[serde(deserialize_with = "i32_from_string")]
    pub time_added: i32,
    #[serde(deserialize_with = "i32_from_string")]
    pub time_updated: i32,
    #[serde(deserialize_with = "i32_from_string")]
    pub time_read: i32,
    #[serde(deserialize_with = "i32_from_string")]
    pub time_favorited: i32,
    pub sort_id: i32,
    pub resolved_title: String,
    pub resolved_url: String,
    pub excerpt: String,
    #[serde(deserialize_with = "bool_from_string")]
    pub is_article: bool,
    #[serde(deserialize_with = "bool_from_string")]
    pub is_index: bool,
    pub has_video: HasVideo,
    pub has_image: HasImage,
    #[serde(deserialize_with = "i32_from_string")]
    pub word_count: i32,
    pub lang: String,
    pub time_to_read: i32,
    pub top_image_url: Option<String>,
    pub listen_duration_estimate: i32,
}

#[derive(Deserialize)]
#[repr(i32)]
pub enum ItemStatus {
    #[serde(rename = "0")]
    Unread = 0,
    #[serde(rename = "1")]
    Archived = 1,
    #[serde(rename = "2")]
    Deleted = 2,
}

#[derive(Deserialize)]
#[repr(i32)]
pub enum HasVideo {
    #[serde(rename = "0")]
    No = 0,
    #[serde(rename = "1")]
    Yes = 1,
    #[serde(rename = "2")]
    IsVideo = 2,
}

#[derive(Deserialize)]
#[repr(i32)]
pub enum HasImage {
    #[serde(rename = "0")]
    No = 0,
    #[serde(rename = "1")]
    Yes = 1,
    #[serde(rename = "2")]
    IsImage = 2,
}
