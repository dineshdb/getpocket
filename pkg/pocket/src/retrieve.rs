use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use util::der::bool_from_number;
use util::ser::serialize_option_bool_as_int;

use super::item::Item;

#[derive(Serialize)]
pub enum State {
    #[serde(rename = "unread")]
    Unread,
    #[serde(rename = "archive")]
    Archive,
    #[serde(rename = "all")]
    All,
}

#[derive(Serialize)]
pub enum DetailType {
    #[serde(rename = "simple")]
    Simple,
    #[serde(rename = "complete")]
    Complete,
}

#[derive(Serialize)]
pub enum Tag {
    #[serde(rename = "_untagged_")]
    Untagged,
    Value(String),
}

#[derive(Serialize)]
pub enum ContentType {
    #[serde(rename = "article")]
    Article,
    #[serde(rename = "video")]
    Video,
    #[serde(rename = "image")]
    Image,
}

#[derive(Serialize)]
pub enum SortBy {
    #[serde(rename = "newest")]
    Newest,
    #[serde(rename = "oldest")]
    Oldest,
    #[serde(rename = "title")]
    Title,
    #[serde(rename = "site")]
    Site,
}

#[derive(Deserialize)]
pub struct RetrieveResponse {
    pub status: i32,
    pub error: Option<String>,
    #[serde(deserialize_with = "bool_from_number")]
    pub complete: bool,
    pub since: i32,
    pub list: HashMap<String, Item>,
}

#[derive(Serialize)]
pub struct GetOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<State>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(serialize_with = "serialize_option_bool_as_int")]
    pub favorite: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tag: Option<Tag>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_type: Option<ContentType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort: Option<SortBy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail_type: Option<DetailType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub search: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub domain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub since: Option<String>,
    pub count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<i32>,
}

impl Default for GetOptions {
    fn default() -> GetOptions {
        GetOptions {
            state: Some(State::Unread),
            favorite: None,
            tag: None,
            content_type: None,
            sort: None,
            detail_type: None,
            search: None,
            domain: None,
            since: None,
            count: 10,
            offset: None,
        }
    }
}
