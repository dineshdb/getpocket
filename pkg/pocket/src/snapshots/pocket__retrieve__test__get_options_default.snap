---
source: pkg/pocket/src/retrieve.rs
expression: options
---
{
  "state": "unread",
  "count": 10
}
